ramips: add support for Apexera NetBot and WiBot

pengphei opened the pull request:
The Apexera NetBot and WiBot is base on the MT7620A SoC.

Apexera WiBot Specification:

- CPU: MT7620A
- DRAM: 256 MiB
- NOR: 16 MiB
- Wireless: 2.4Ghz
- Ethernet speed: 10/100
- Ethernet ports: 1
- 1x USB Hub (with 3 ports)
  - 1x USB Converted Mini PCIe Slot (for PCIE 4G Adapter)
  - 1x USB 3.0
  - 1x USB Power Output: DC 5V Max 1A
  - 1x USB Disk with 8GiB (on the board)
- 1x Sim Card Slot(with pcie slot)
- 1x Micro USB Power Input: 5V Max 2.5A
- 1x microSD reader
- Debug Serial: 57600,8N1
- Battery: 6000mAh/3.8V
- STM32F0: stm32f030f4p6 charging manage chip.
- Reset Button
- LED: In Charge lEDs(4 streaming LED), Power, Wifi, 4G available.

Apexera NetBot Specification:

- CPU: MT7620A
- DRAM: 256 MiB
- NOR: 16 MiB
- Nand: 8GiB (as usb nand disk)
- Wireless: 2.4Ghz
- Ethernet speed: 10/100
- Ethernet ports: 1
- 4x USB 2.0
- 1x Micro USB Power Input: 5V Max 2.5A
- 1x microSD reader
- Debug Serial: 57600,8N1
- Reset Button
- I2S Output
- LED: wifi, usb, ethernet

The PCIE slot of WiBot is designed for 4G dongle with sim card slot. Customers could
add the 4G dongle prefered.

Installation:
Apply factory image via http web-gui with sysupgrade.bin.

Signed-off-by: Han Pengfei <pengphei@foxmail.com>

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

pengphei committed:
ramips: add support for Apexera NetBot

The Apexera NetBot is based on the MT7620A SoC.

Apexera NetBot Specification:

- CPU: MT7620A
- DRAM: 256 MiB
- NOR: 16 MiB
- Nand: 8GiB (as usb nand disk)
- Wireless: 2.4Ghz
- Ethernet speed: 10/100
- Ethernet ports: 1
- 4x USB 2.0
- 1x Micro USB Power Input: 5V Max 2.5A
- 1x microSD reader
- Debug Serial: 57600,8N1
- Reset Button
- I2S Output
- LED: wifi, usb, ethernet

Installation:
Apply factory image via http web-gui with sysupgrade.bin.

Signed-off-by: Han Pengfei <pengphei@foxmail.com>

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&factory 0>;
```
Consider using 0x0 for consistency.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_wibot.dts:
```diff
@@ -0,0 +1,203 @@
+};
+
+&wmac {
+	ralink,mtd-eeprom = <&factory 0>;
```
Consider using 0x0 for consistency.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -73,6 +73,31 @@ define Device/amit_jboot
 endef
 
+define Device/apexera-netbot
+  DTS := ApexeraNetBot
```
Remove this line.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -73,6 +73,31 @@ define Device/amit_jboot
+TARGET_DEVICES += apexera-netbot
+
+define Device/apexera-wibot
+  DTS := ApexeraWiBot
```
Remove this line.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -73,6 +73,31 @@ define Device/amit_jboot
   DEVICE_PACKAGES := jboot-tools kmod-usb2 kmod-usb-ohci
 endef
 
+define Device/apexera-netbot
```
"apexera_netbot" instead of "apexera-netbot".
Did you build-test, because that should have caused an error?

pengphei replied:
I figured out I have tested the commit maybe one or two week ago before we have made the new rule for device image building and dts binding with it.

pengphei replied:
For now, we don't have to define the DTS variable with image building, and the dts file name is defined with the target name, right?

adrianschmutzler replied:
The DTS file name is constructed from MTK_SOC and the device definition name:
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/image/Makefile#L28

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -73,6 +73,31 @@ define Device/amit_jboot
+       kmod-usb-ohci kmod-sdhci-mt7620 kmod-usb-storage \
+       kmod-usb-storage-extras kmod-usb-storage-uas
+endef
+TARGET_DEVICES += apexera-netbot
```
Also underscore here.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -73,6 +73,31 @@ define Device/amit_jboot
+define Device/apexera-netbot
+  DTS := ApexeraNetBot
+  IMAGE_SIZE := 16064k
+  DEVICE_TITLE := Apexera NetBot
```
Use new scheme with DEVICE_VENDOR and DEVICE_MODEL.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -73,6 +73,31 @@ define Device/amit_jboot
+endef
+TARGET_DEVICES += apexera-netbot
+
+define Device/apexera-wibot
```
Underscore.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -73,6 +73,31 @@ define Device/amit_jboot
+define Device/apexera-wibot
+  DTS := ApexeraWiBot
+  IMAGE_SIZE := 16064k
+  DEVICE_TITLE := Apexera WiBot
```
DEVICE_VENDOR etc.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -73,6 +73,31 @@ define Device/amit_jboot
+       kmod-usb-ledtrig-usbport kmod-usb-storage \
+       kmod-usb-storage-extras kmod-usb-storage-uas
+endef
+TARGET_DEVICES += apexera-wibot
```
Underscore.

adrianschmutzler commented:
I think both devices still have enough in common to create a shared DTSI.

adrianschmutzler commented:
Which of the interfaces bears the MAC address on the device's label/sticker/case?

pengphei commented:
> I think both devices still have enough in common to create a shared DTSI.

hmmm, these two devices are totally different. I have no idea how to share the dtsi.

netbot is targeted at mutiple usb slots and worked as a common wireless router.
wibot is targeted at 3g wireless and battery powered mobile router.

adrianschmutzler commented:
Well, one could share:
- console
- pcf8563
- several of the nodes just setting status = okay
- whole spi block
- wmac

But if you feel like that's nonsense because the device are too different, I won't force you.

adrianschmutzler commented:
... But if the device are that different, I will make sense to add them in separate commits.

pengphei commented:
> ... But if the device are that different, I will make sense to add them in separate commits.

ok. I will try to make seperate pull requirest

adrianschmutzler commented:
Please do not open a new Pull Request, just have two commits in this one.

pengphei committed:
ramips: add support for Apexera WiBot

The Apexera WiBot is based on the MT7620A SoC.

Apexera WiBot Specification:

- CPU: MT7620A
- DRAM: 256 MiB
- NOR: 16 MiB
- Wireless: 2.4Ghz
- Ethernet speed: 10/100
- Ethernet ports: 1
- 1x USB Hub (with 3 ports)
  - 1x USB Converted Mini PCIe Slot (for PCIE 4G Adapter)
  - 1x USB 3.0
  - 1x USB Power Output: DC 5V Max 1A
  - 1x USB Disk with 8GiB (on the board)
- 1x Sim Card Slot(with pcie slot)
- 1x Micro USB Power Input: 5V Max 2.5A
- 1x microSD reader
- Debug Serial: 57600,8N1
- Battery: 6000mAh/3.8V
- STM32F0: stm32f030f4p6 charging manage chip.
- Reset Button
- LED: In Charge lEDs(4 streaming LED), Power, Wifi, 4G available.

The PCIE slot of WiBot is designed for 4G dongle with sim card slot. Customers could
add the 4G dongle prefered.

Installation:
Apply factory image via http web-gui with sysupgrade.bin.

Signed-off-by: Han Pengfei <pengphei@foxmail.com>

pengphei commented:
@adrianschmutzler 

I have fixed all the unresolved issues. and seperate netbot and wibot as two commits. please review again. thanks. :)

adrianschmutzler commented:
@pengphei All of my comments have been adressed, as it looks.
However, I cannot provide a deep review of these devices, so you will have to wait for someone else to do that.

pengphei commented:
> @pengphei All of my comments have been adressed, as it looks. :-)
> However, I cannot provide a deep review of these devices, so you will have to wait for someone else to do that.

I have got the SCH from the vendor, https://github.com/apexera/mars-moon-guide/blob/master/source/_static/apexera-wibot-v1-51.pdf , I hope it will help. 

And I think, this device is on sale, but not that widely used as router.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_wibot.dts:
```diff
@@ -0,0 +1,186 @@
+
+		wifi {
+			label = "wibot:white:wifi";
+			gpios = <&gpio1 4 GPIO_ACTIVE_HIGH>;
```
No trigger?

pengphei replied:
I have checked the dts referring to mt7620a, as for wifi led, no trigger found for it. do you have any suggestion ?

adrianschmutzler replied:
And it doesn't work if you just put "linux,default-trigger = "phy0tpt";" there?

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+
+	aliases {
+		led-boot = &led_wifi;
+		led-failsafe = &led_wifi;
```
It is discouraged to hijack LEDs with other purposes here. If you only have LEDs which would be ambiguous on a running device, remove the aliases section.

pengphei replied:
I have read the SCH of this device, And it has only one led that could be used as the alias function, and I think is necessary to tell the user which state it's working, so I choose this.

adrianschmutzler replied:
Still, I do not like this ambiguity, please remove the aliases in order to move this forward.

pengphei commented:
@adrianschmutzler I have connected the vendor, and they are very kindly to donate several devices. and you know who is in charge of this?

adrianschmutzler commented:
@blocktrron 

Maybe have a look at this and the most recent comment from pengphei.

pengphei commented:
> @blocktrron
> 
> Maybe have a look at this and the most recent comment from pengphei.

@adrianschmutzler  what should I do for the next step?

adrianschmutzler commented:
> what should I do for the next step?

You will have to wait. I'm not a core-developer, so I can't merge your code.
For the open issues, it would be good to have at least a second opinion anyway.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -73,6 +73,17 @@ define Device/amit_jboot
 endef
 
+define Device/apexera_netbot
+  MTK_SOC := mt7620a
```
Rename to "SOC" due to 58e71e322bc0e0edd08157075e9022c15ae33ec5.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+		led-failsafe = &led_wifi;
+	};
+
+	gpio-leds {
```
Should be "leds"

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+		compatible = "gpio-leds";
+
+		led_wifi: wifi {
+			label = "netbot:led:wifi";
```
Replace ":led:" by the appropriate color.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+		};
+
+		wan {
+			label = "netbot:led:wan";
```
Replace ":led:" by the appropriate color.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+			gpios = <&gpio2 26 GPIO_ACTIVE_HIGH>;
+		};
+
+		usb2g {
```
usb2_green

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+			gpios = <&gpio2 1 GPIO_ACTIVE_HIGH>;
+		};
+
+		usb2r {
```
usb2_red and accordingly below

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Please check whether this can be increased.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+
+&ethernet {
+	mtd-mac-address = <&factory 0x10>;
+	ralink,port-map = "wllll";
```
Do you really need this for a one-port?

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x10>;
```
This can't be correct of you read the caldata from 0x0.

Please provide a list of MAC addresses on stock FW (2.4 Ghz, 5 Ghz, ethernet) and a hexdump of the factory partition. For further reading: https://openwrt.org/docs/guide-developer/mac.address

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -73,6 +73,17 @@ define Device/amit_jboot
+  IMAGE_SIZE := 16064k
+  DEVICE_PACKAGES := kmod-i2c-ralink kmod-rtc-pcf8563 kmod-sound-core \
+       kmod-sound-mt7620 kmod-usb2 kmod-usb-ohci kmod-sdhci-mt7620 \
+       kmod-usb-storage kmod-usb-storage-extras kmod-usb-storage-uas
```
You can at least remove kmod-sound-core and kmod-usb-storage here.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+		i2s_uartf {
+			ralink,group = "uartf";
+			ralink,function = "i2s uartf";
+		};
```
What does this block?

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_netbot.dts:
```diff
@@ -0,0 +1,177 @@
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
```
state_default already exists in parent DTSI, so you can use it directly, e.g.
https://github.com/openwrt/openwrt/blob/38bdfcdb87e9b3a1232e9820bd6534da4c28282b/target/linux/ramips/dts/mt7620a_iptime_a104ns.dts#L54

adrianschmutzler commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -144,6 +144,7 @@ ramips_setup_interfaces()
 	allnet,all0256n-8m|\
 	allnet,all5002|\
 	allnet,all5003|\
+	apexera,netbot|\
```
Can you explain why you need this:
		ucidef_add_switch "switch0"
		ucidef_add_switch_attr "switch0" "enable" "false"
?

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_wibot.dts:
```diff
@@ -0,0 +1,186 @@
+		led-upgrade = &led_status;
+	};
+
+	gpio-leds {
```
-> leds

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_wibot.dts:
```diff
@@ -0,0 +1,186 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Raise frequency?

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_wibot.dts:
```diff
@@ -0,0 +1,186 @@
+
+&ethernet {
+	mtd-mac-address = <&factory 0x10>;
+	ralink,port-map = "llllw";
```
port-map for one-port?

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_wibot.dts:
```diff
@@ -0,0 +1,186 @@
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x10>;
```
See other device. that would be inside the wifi caldata.

adrianschmutzler commented on target/linux/ramips/dts/mt7620a_apexera_wibot.dts:
```diff
@@ -0,0 +1,186 @@
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
```
Use state_default.

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -84,6 +84,18 @@ define Device/apexera_netbot
 TARGET_DEVICES += apexera_netbot
 
+define Device/apexera_wibot
+  MTK_SOC := mt7620a
```
SOC

adrianschmutzler commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -84,6 +84,18 @@ define Device/apexera_netbot
+  DEVICE_MODEL := WiBot
+  IMAGE_SIZE := 16064k
+  DEVICE_PACKAGES := kmod-i2c-ralink kmod-rtc-pcf8563 kmod-sdhci-mt7620 \
+       kmod-usb2 kmod-usb-ohci kmod-usb-ledtrig-usbport kmod-usb-storage \
```
remove usb-storage, it should be pulled by the others.

adrianschmutzler commented:
Ping.

adrianschmutzler commented:
There haven't been any updates for quite some time, so I'm closing this now.

Feel free to ask for reopening when this is continued. Please note that this has to be reopened _before_ pushing to the GitHub branch.
